### DISCUSSION TITLE
Were does pyusb define its error codes?

### DISCUSSION THREAD

**Discussion by brianreinhold** [pyusb/pyusb#445] [Q&A] [answered]
If I unplug a connected device whole transferring data I get a ```USBError``` with ```errno = 19```. As far as I can tell from the message along with the error that is the same as ```LIBUSB_ERROR_NO_DEVICE:'No such device (it may have been disconnected)',``` where
```LIBUSB_ERROR_NO_DEVICE = -4``` .

How does -4 get converted to 19?  Does one simply add 23 to all the LIBUSB error codes to get ```errno```?

What are the ```errno``` for these guys?

LIBUSB_ERROR_IO = -1
LIBUSB_ERROR_INVALID_PARAM = -2
LIBUSB_ERROR_ACCESS = -3
LIBUSB_ERROR_NO_DEVICE = -4
LIBUSB_ERROR_NOT_FOUND = -5
LIBUSB_ERROR_BUSY = -6
LIBUSB_ERROR_TIMEOUT = -7
LIBUSB_ERROR_OVERFLOW = -8
LIBUSB_ERROR_PIPE = -9
LIBUSB_ERROR_INTERRUPTED = -10
LIBUSB_ERROR_NO_MEM = -11
LIBUSB_ERROR_NOT_SUPPORTED = -12
LIBUSB_ERROR_OTHER = -99

How does ```errno``` relate to ```backend_error_code```

Thanks for anyone who can deobfuscate this for me!

**Comment by jonasmalacofilho** [comment]
https://github.com/pyusb/pyusb/blob/777dea9d718e70d7323c821d4497c706b35742da/usb/backend/libusb1.py#L119-L135

**Reply by brianreinhold** [reply]
The question remains. I see the LIBUSB_ERROR_* values, but I don't see a table of errno values. They are not the same for the same error. In the USBErr exception I get errno, not LIBUSB_ERROR_*. If I got the latter, I would be much happier as I know where to find those values.

**Reply by jonasmalacofilho** [reply]
> The question remains.

I'm sorry, I must be misunderstanding your question somehow...

Don't you want to know where in the PyUSB source code is the mapping  `LIBUSB_ERROR_* -> errno`?

**Reply by jonasmalacofilho** [reply]
If you say "the question remains", then please answer the follow up reply,[^1] even if just to acknowledge that you got that the original reply did point to what had asked about... or, if there's misunderstanding on my part, then please clarify the question.

[^1]: FYI, if this became a pattern, I would eventually stop answering your questions.

**Reply by brianreinhold** [reply]
I think I need the reverse. I have code like this:
```
        except usb.core.USBError as e:
            print(f'USB error raised with code {e.backend_error_code}')
            print(e)
            if e.errno == 19:   # This is error 'No such device (it may have been disconnected)'
                                # TODO find the errno for a time out
                self.ep_in = None
                self.ep_out = None
            return False
```
I get the 'modified' error code. I was looking for a list of these possible errno values and what they mean. So go from errno to meaning.

```
errno          meaning

0                   ?
1                   ?
...
18                 ?
19                No such device
20                ?
```

**Reply by jonasmalacofilho** [reply, marked as answer]
That's recorded in `usb.core.USBError`: the `USBError.backend_error_code` field points to the original backend error value, which you can then match against what that backend may return.

Possibly, but not only, by comparing against, for example, `usb.backend.libusb1.LIBUSB_ERROR_NO_DEVICE `.

---

In the case of timeout errors, and with a recent enough version of PyUSB, you can directly catch the `usb.core.USBTimeoutError`.
